### ISSUE TITLE
Clarify Kafka transport mapping of "content-type"

### ISSUE THREAD

**Issue by gunnarmorling** [cloudevents/spec#530]
Structured mode is described as a way for conveying messages completely self-contained, without relying on any Kafka headers. The example in [3.3.4](https://github.com/cloudevents/spec/blob/master/kafka-transport-binding.md#334-example) shows this as a header, though. I'll be happy to send a PR for doing the needed adjustments, but it's not quite clear to me yet what it should be.

**Comment by duglin** [comment]
@bluemonk3y @clemensv 
Is the example ok because `content-type` is special like it is in http?

I'm actually more wondering about this line:
```
"datacontenttype" : "com.example.someevent",
```
shouldn't that be a valid type per RFC 2046  ?

**Comment by gunnarmorling** [comment]
Yes, so there's both in this example, `datacontenttype` and `content-type`. The relationship between the two is not quite clear to me, I think this warrants some clarification. 

Also 3.3.1 speaks of "The [Kafka content-type] property field"; apart from the odd formatting with the brackets I'm not quite clear what a "property field" in the context of Kafka actually is. If it's indeed meant to be used as a message header (as indicated example, but a bit against the spirit of structured type), then I think it should be named like that.

**Comment by bluemonk3y** [comment]
My interpretation is:
* `type`: represents the class or data type itself; the deserialized format of the class. For example if you decode a payload then a `com.company.Payment` is the data type.

* `datacontenttype` represents the serialization format for example, `application/avro` json etc

The AMQP spec uses type and  content-type respectively
https://github.com/cloudevents/spec/blob/master/amqp-transport-binding.md

However, it also seems like the 2 might be confused with each other (I had to look at other docs to confirm the above points which is a fail IMO ;)  Would `contentformat` or similar make more sense? It seems like type being used in the 2 attributes makes it confusing.

**Comment by gunnarmorling** [comment]
@bluemonk3y, ok, so it actually is the case then that in _structured_ mode a Kafka message-level header should be used to convey the serialization format in RFC 2046 format?

If so, the followin sentence is confusing to me:

> The binary mode only applies to Kafka 0.11.0.0 and above, because Kafka 0.10.x.x and below lack support for message level headers.

The same applies to structured mode then, too, as otherwise the fact it's only given for _binary_ mode to me implies that structured would work with earlier versions. So I think this sentence should be removed then and be replaced by something stating that 0.11.0.0 is needed in both modes due the reliance on message headers.

**Comment by gunnarmorling** [comment]
@duglin, @bluemonk3y, so I've been giving it a shot in PR https://github.com/cloudevents/spec/pulls/534:

- [ ] Made the statement about the Kafka version a general one as it applies to both modes as per the discussion above
- [ ] Replaced `content-type` with `ce_datacontenttype` in the structured mode examples; this is inline with what's said in 3.:

> "The receiver of the event can distinguish between the two content modes by inspecting the ce_datacontenttype Header of the Kafka message. "
- [ ] Clarified it's a message level header
- [ ] Replaced `datacontenttype` within the structured mode example payload with `type`, aligning with AMQP and avoiding the confusion with the content type property

Would that help?

**Comment by bluemonk3y** [comment]
yes that is correct but I cannot see your PR

**Comment by gunnarmorling** [comment]
Doh, seems my link is broken. Here it is:
https://github.com/cloudevents/spec/pull/534

**Comment by bluemonk3y** [comment]
Yes, however in structured mode the headers are avoided and the ce_datacontenttype would be part of the payload; there is not other means of sending the attribute when headers are not supported

**Comment by gunnarmorling** [comment]
> Yes, however in structured mode the headers are avoided

Isn't this in contradiction though with 3.:

> "The receiver of the event can distinguish between the two content modes by inspecting the ce_datacontenttype Header of the Kafka message. "

That is, unless a receiver should conclude from the *absence* of the header that it's structured mode? Also in the example in 3.3.1 it *is* shown as a message header and not as part of the payload. So IMO it needs to be aligned in one of the two ways.

**Comment by bluemonk3y** [comment]
agree - the header shown in 3.3.1 json should be removed and **absence** used to determine structured or binary. It would not make sense to have differing mechanisms for different versions of kafka.

**Comment by gunnarmorling** [comment]
Ok, so I feel there's stil something unclear; when looking at the [JSON event format](https://github.com/cloudevents/spec/blob/master/json-format.md), there it's described how a JSON cloud event structure can take *another* format in the `data` attribute, e.g. XML or Avro. Is this actually supported in the Kafka transport? Or is the assumption that when using structured mode with JSON, then `data` is also JSON and cannot be something else like Avro, as e.g. shown the JSON event format spect itself?

**Comment by gunnarmorling** [comment]
I've updated my PR along that assumption, i.e. `data` in structured mode cannot have be of a different content type than the Kafka message value itself. If this actually *is* meant to be supported, e.g. a Kafka message value that's JSON with a `data` attribute whose content type is Avro, then I think this will need further clarification. A `datacontenttype` attribute *within* the JSON then would probably the way to go, but it's quite confusing then with the similarly `ce_datacontenttype` message header.

**Comment by bluemonk3y** [comment]
>  data in structured mode cannot be of a different content type than the Kafka message value itself.
It should be possible to have a different data format in struct mode (as shown in the json example). 

>A datacontenttype attribute within the JSON then would probably the way to go, but it's quite confusing then with the similarly ce_datacontenttype message header.

the top level json attributes are equivalent to the header attributes. struct uses the top level json attributes and binary uses headers. i.e. there will be one or the other.

Does that help? - I feel like we might be going around in circles ;( 

We could discuss on the next cncf call or jump on a zoom on Monday? (Im in BST TZ)

**Comment by duglin** [comment]
Please chat (and try to resolve) before eod tuesday since that's the deadline for updates for PR to be approved this week - and we'd like to close all v1.0 PRs this week so we can start a vote!!! woo hoo!

**Comment by gunnarmorling** [comment]
> Does that help? - I feel like we might be going around in circles ;(

Yes, sorry about that. On the bright side, if we do some changes so I get it, chances are it will be better for others reading the spec in the future, too.

I would love to do a call, but I'm on a flight to the US on Monday, so this will be tricky. I have a 1:30 stop-over at 11:10 BST, so I *might* be able to do a quick call at say 11:30 BST give or take, connectivity provided etc.

In the meantime, perhaps you can take a look at my PR https://github.com/cloudevents/spec/pull/534? This already might clarify most things from my PoV.

Besides these changes, my remaining question is the following. Consider this example CE from the JSON event format spec:

```
{
    "specversion" : "1.0-rc1",
    "type" : "com.example.someevent",
    "source" : "/mycontext",
    "id" : "A234-1234-1234",
    "time" : "2018-04-05T17:31:00Z",
    "comexampleextension1" : "value",
    "comexampleothervalue" : 5,
    "datacontenttype" : "text/xml",
    "data" : "<much wow=\"xml\"/>"
}
```

How would this be mapped to structured mode in the Kafka transport? Will the `ce_datacontenttype` Kafka message header be set to `application/cloudevents+json` (or omitted, resulting to the same value being applied by default) and the Kafka message value will be the same JSON as above?

Then it's clear to me, a consumer would

* first have to examine the `ce_datacontenttype` Kafka message header to conclude the Kafka message value is JSON
* then JSON-parse this value, read the `datacontenttype` JSON value to learn that the `data` JSON value must be parsed as XML

**Comment by bluemonk3y** [comment]
@gunnarmorling  - I sent you a twitter direct message so we can try to sync on Monday. The PR looks fine except the underlying problem is that struct applies to versions of Kafka that may or may not support headers.

**Comment by gunnarmorling** [comment]
> The PR looks fine except the underlying problem is that struct applies to versions of Kafka that may or may not support headers.

Yes, so I tried to address that by clarifying that the `ce_datacontenttype` header is optional and that a consumer must handle it as JSON if the header is not present. I.e. on Kafka versions without the header, only structured mode with JSON could be used.

**Comment by bluemonk3y** [comment]
@clemensv  - it would be good to get your take on this. Kafka 2.4 only recently provided access to Kafka headers (October). This means the default format for events from Kafka Connect would be struct. At the moment struct is json and ideally

**Comment by gunnarmorling** [comment]
> Kafka 2.4 (oct) only recently provided access to Kafka headers (October)
>  in Connect (part of the kafka eco system). This means the default format
> for events from Kafka Connect would be struct.

To clarify on that, actual Kafka Connect *connectors* could also write
headers before 2.4. But *converters*  will only have that ability in 2.4. I
was bringing this up, as a generic Kafka Connect converter that produces
CloudEvents may be a reasonable implementation for bringing CE support to
the Kafka Connect ecosystem.

Am Di., 15. Okt. 2019 um 14:37 Uhr schrieb Neil Avery <
notifications@github.com>:

**Comment by duglin** [comment]
> Structured mode is described as a way for conveying messages completely self-contained, without relying on any Kafka headers. The example in [3.3.4](https://github.com/cloudevents/spec/blob/master/kafka-protocol-binding.md#334-example) shows this as a header, though. I'll be happy to send a PR for doing the needed adjustments, but it's not quite clear to me yet what it should be.

I think the example is correct w.r.t. the content-type. datacontenttype should have a mime type value, or be renamed to `type` as you did in your PR.

Both content-type and datacontenttype CAN appear in the same message when structured mode is used. content-type will be a protocol header while datacontenttype will be in the payload and will describe the format of the `data` attribute. In binary mode I don't think we'll ever have content-type and datacontenttype appear at the same time since content-type should be all that's needed as it describes the format of the payload. datacontenttype woudl be redundant at that point.

Does this help?

**Comment by duglin** [comment]
Net I think:
- 3.3.4 needs to be updated to make datacontenttype have a mime type, and for completeness we should add a `type` attribute
- the references to `ce_datacontentype` should be `content-type` - 4 of them

But I'm not too familiar with Kafka - @clemensv, does this sound right?

**Comment by alanconway** [comment]
The analogy from AMQP is:
1. amqp content-type is always set and at the AMQP level it has the standard meaning - it describes the formatting of the AMQP data section as a media-type.
2. In a *structured* message AMQP content-type is set to the structured CE format, e.g. application/cloudevents+json. Decoding such an event is a 2 stage process: first we look at AMQP content-type and discover we have a JSON CE event. Then we decode the JSON event independently according to the JSON CE rules and discover it internally has a (possibly different) datacontenttype and data.
3. In a *binary* message AMQP content-type serves 2 purposes: it describes the AMQP data section *and* is the CE datacontenttype. This is because in binary encoding the CE data and the AMQP data section are exactly the same thing.
